### PR TITLE
Added support for dynamicTemplateData

### LINKED ISSUE #689: Support for dynamic_template_data
#### Issue Summary
When is the support coming for the usage of `dynamic_template_data` as per the documentation in https://github.com/sendgrid/email-templates/tree/master/dynamic-templates

Looks like the go library already supports it: https://github.com/sendgrid/smtpapi-go/blob/master/smtpapi.go#L27

### PR DESCRIPTION
**Pull request by mvpspl619**
# Fixes # 
Fixes #689 

### Checklist
- [x] I have made a material change to the repo (functionality, testing, spelling, grammar)
- [x] I have read the [Contribution Guide] and my PR follows them.
- [x] I updated my branch with the master branch.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation about the functionality in the appropriate .md file
- [x] I have added in line documentation to the code I modified

### Short description of what this PR does:
- Adds support for SendGrid's latest feature for enabling handlebars syntax in transactional templates
- Checks the `templateId`, and decides whether it is dynamic template or not. (Dynamic template ids start with `d-`. 
- Does not send `substitutions` if the `templateId` is for a dynamic template
- Accepts `dynamicTemplateData` in the mail object directly and merges with `dynamicTemplateData` in each `personalization` object

### CODE REVIEW

**Comment by SendGridDX** [comment]
[CLA assistant check](https://cla.sendgrid.com/sendgrid/sendgrid-nodejs?pullRequest=691) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.sendgrid.com/sendgrid/sendgrid-nodejs?pullRequest=691) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.sendgrid.com/check/sendgrid/sendgrid-nodejs?pullRequest=691) it.</sub>

**Comment by mvpspl619** [comment]
@thinkingserious As we're on a time crunch, we published the new packages with this version to our private npm org and we'll be using it until this PR gets reviewed/merged. Hope its okay.

File: packages/helpers/classes/mail.js [hunk 1]
```diff
@@ -247,7 +273,7 @@ class Mail {
     ) {
       throw new Error('Provide at least one of to, cc or bcc');
     }
-    this.addPersonalization(new Personalization({to, cc, bcc}));
```
Comment (adamreisnz) [inline packages/helpers/classes/mail.js]: The code style in this package doesn't use leading/trailing spaces after/before curly brackets, could you revert this change?

File: packages/helpers/classes/mail.js [hunk 2]
```diff
@@ -21,6 +21,7 @@ class Mail {
   constructor(data) {
 
     //Initialize array and object properties
+    this._isDynamic = false;
```
Comment (adamreisnz) [inline packages/helpers/classes/mail.js]: I'm thinking just us `this.isDynamic` for the flag, e.g. without the underscore.

File: packages/helpers/classes/mail.js [hunk 3]
```diff
@@ -288,6 +314,28 @@ class Mail {
...
+   * Helper which applies globally set dynamic_template_data to personalizations
+   */
+  applyDynamicTemplateData(personalization) {
+    if (personalization instanceof Personalization) {
+      personalization.reverseMergeDynamicTemplateData(this.dynamicTemplateData);
+    }
+  }
+
+  /**
+   * Set substitutions
```
Comment (adamreisnz) [inline packages/helpers/classes/mail.js]: Fix comment to avoid confusion

File: packages/helpers/classes/mail.spec.js [hunk 4]
```diff
@@ -8,10 +8,10 @@ const Mail = require('./mail');
 /**
  * Tests
  */
-describe('Mail', function() {
+describe('Mail', function () {
```
Comment (adamreisnz) [inline packages/helpers/classes/mail.spec.js]: Can you also undo these style corrections? Please respect the code style as indicated by the editor config and eslint config. If you enable ESLint in your editor it should automatically fix these issues.

File: packages/helpers/classes/personalization.js [hunk 5]
```diff
@@ -271,6 +274,34 @@ class Personalization {
     this.substitutionWrappers = wrappers;
   }
 
+  /**
+   * Reverse merge dynamic template data, preserving existing ones
+   */
+  reverseMergeDynamicTemplateData(dynamicTemplateData) {
```
Comment (adamreisnz) [inline packages/helpers/classes/personalization.js]: Can we call this method `deepMergeDynamicTemplateData` instead? I think that's what's happening here, so that description would be a bit clearer and a better fit imo.

File: packages/helpers/classes/personalization.js [hunk 6]
```diff
@@ -279,11 +310,11 @@ class Personalization {
     //Get data from self
     const {
       to, cc, bcc, subject, headers, customArgs, sendAt,
-      substitutions, substitutionWrappers,
+      substitutions, substitutionWrappers, dynamicTemplateData,
     } = this;
 
     //Initialize with mandatory values
-    const json = {to};
+    const json = { to };
```
Comment (adamreisnz) [inline packages/helpers/classes/personalization.js]: Code style

File: packages/helpers/helpers/merge-data-deep.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
+'use strict';
```
Comment (adamreisnz) [inline packages/helpers/helpers/merge-data-deep.js]: Can we use the existing and proven package [deepmerge](https://www.npmjs.com/package/deepmerge) for this instead of rolling our own helper? From experience, this can be a tricky beast to get right, so I'd rather use an existing package which is being downloaded and used a lot, rather than try to replicate and have to maintain this part of the logic ourselves.

In particular merging of arrays can be tricky, which this package handles very flexibly.

**Comment by adamreisnz** [comment]
@ben-grid Looks good to me, just a few tweaks that I'd recommend, as well as using the external library `deepmerge` instead of making our own.

If this gets merged I'd probably pick up the fix from #687 as well, as it touches the same code.

**Comment by ben-grid** [comment]
@adamreisnz @mvpspl619 You both friggin rock! Thanks so much for your contributions here!

**Comment by mvpspl619** [comment]
@adamreisnz Can you take a look one more time? Sorry about the delay, I couldn't figure out why my VSCode kept overriding the trailingSpaces config, and then realized it was in my userSettings, so turned it off and voila.

**Comment by thinkingserious** [comment]
Hello @mvpspl619,<br /><br/>Thanks again for the PR!<br />                                              <br />We want to show our appreciation by sending                                              you some swag. Could you please fill out                                              <a href="https://dx.sendgrid.com/swag">this form</a> so we can send it to you? Thanks!                                             <br /><br />Team SendGrid DX

**Comment by ron23** [comment]
Hey, may I ask what is dynamic templating and what's the difference between that and the standard "Transactional Templating"? 
I also don't get how to create one. If I create one, I automatically get an ID, which doesn't start with a "d-" - does that mean it's not "dynamic"?

**Comment by adamreisnz** [comment]
@ron23 yes I think that's correct. See https://github.com/sendgrid/sendgrid-nodejs/issues/221#issuecomment-389029979 for more information on what dynamic templates are (going to be). Essentially it adds support for handlebars style templates with additional templating language syntax, like loops and conditional statements etc.

**Comment by ben-grid** [comment]
@ron23 Adam nailed it, I just wanted to add, we are running a beta of this right now, which is why/how Adam and Raju have access to this functionality and you don't yet. If you'd like to participate just drop me an email @ ben@sendgrid.com and include your username. Cheers!

**Comment by isaachinman** [comment]
I cannot believe how long it took me to find this buried in the release notes. This info needs to be in the main READMEs. These dynamic transactional templates are apparently the default now, and I lost almost two hours tracking this down.

**Comment by nemoeric** [comment]
I lost 3 hours looking for this `dynamicTemplateData` parameter :/
I do agree with isaachinman here, this should be visible more easily as Sendgrid now asks its user to migrate to Transactionnal vs Legacy ... !

**Comment by isaachinman** [comment]
It's great work from @mvpspl619, but it's pretty unfortunate that the maintainers here apparently do not require front-facing documentation to accompany new features. I'm happy to submit a PR which updates the docs if it will be accepted.

Is there an explanation why `substitutions` cannot still be used, even with the new dynamic templates? Why do we need a new `dynamicTemplateData` if apparently a check is still being run for this `d-` pattern? @nemoeric and I, for example, would not even be here if this functionality worked without breaking changes to the API.

**Comment by thinkingserious** [comment]
FYI, here is a repost of the announcement made this morning:

Hey all! Great news––

### Dynamic content for Transactional Templates is here!
This morning, our team publicly launched dynamic content for transactional templates. It is now available for all customers sending over v3 of our Mail Send API. Iterate over lists, handle conditionals and more, thanks to native support for a subset of Handlebars syntax!

More information can be found in our [blog post announcement](https://sendgrid.com/blog/how-to-use-sendgrids-dynamic-templates-for-your-transactional-emails/).

**Documentation:** 
- [How to Send an Email With Dynamic Transactional Templates](https://sendgrid.com/docs/User_Guide/Transactional_Templates/how_to_send_an_email_with_transactional_templates.html)
- [Using Handlebars](https://sendgrid.com/docs/User_Guide/Transactional_Templates/Using_handlebars.html)
- [API Docs](https://dynamic-templates.api-docs.io/3.0/mail-send-with-dynamic-transactional-templates/v3-mail-send)

Thank you, on behalf of the entire SendGrid team, for all of the feedback throughout the beta program. 🙌🏻

**Comment by thinkingserious** [comment]
@isaachinman @nemoeric,

My apologies for the poor experience. I've created [this issue](https://github.com/sendgrid/sendgrid-nodejs/issues/703) and added it to our backlog to improve usability. Please feel free to add to that issue if you feel I missed anything.

Thanks!

With Best Regards,

Elmer
